code/modules/mob/living/carbon/xenomorph/mutation_datum.dm

```diff
@@ -0,0 +1,167 @@
+GLOBAL_DATUM_INIT(mutation_selector, /datum/mutation_datum, new)
+
+/datum/mutation_datum
+	interaction_flags = INTERACT_UI_INTERACT
+
+/datum/mutation_datum/ui_state(mob/user)
+	return GLOB.hive_ui_state // Similar purpose.
+
+/datum/mutation_datum/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "MutationSelector", "Mutation Selector")
+		ui.open()
+
+/datum/mutation_datum/ui_data(mob/living/carbon/xenomorph/xeno_user)
+	var/list/data = list()
+
+	data["shell_chambers"] = length(xeno_user.hive?.shell_chambers)
+	data["spur_chambers"] = length(xeno_user.hive?.spur_chambers)
+	data["veil_chambers"] = length(xeno_user.hive?.veil_chambers)
+	data["biomass"] = xeno_user.biomass
+	switch(xeno_user.xeno_caste.tier)
+		if(XENO_TIER_ONE)
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T1
+		if(XENO_TIER_TWO)
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T2
+		if(XENO_TIER_THREE)
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T3
+		else
+			data["cost"] = XENO_UPGRADE_BIOMASS_COST_T4
+
+	var/list/upgrades = list(
+		list(
+			"category" = "Survival",
+			"upgrades" = list(
+				list(
+					"name" = "Carapace",
+					"desc" = "Increase our armor."
+				),
+				list(
+					"name" = "Regeneration",
+					"desc" = "Increase our health regeneration."
+				),
+				list(
+					"name" = "Vampirism",
+					"desc" = "Leech from our attacks."
+				)
+			)
+		),
+		list(
+			"category" = "Attack",
+			"upgrades" = list(
+				list(
+					"name" = "Celerity",
+					"desc" = "Increase our movement speed."
+				),
+				list(
+					"name" = "Adrenaline",
+					"desc" = "Increase our plasma regeneration."
+				),
+				list(
+					"name" = "Crush",
+					"desc" = "Increase our damage to objects."
+				)
+			)
+		),
+		list(
+			"category" = "Utility",
+			"upgrades" = list(
+				list(
+					"name" = "Toxin",
+					"desc" = "Inject toxins into our target."
+				),
+				list(
+					"name" = "Pheromones",
+					"desc" = "Ability to emit pheromones."
+				),
+				list(
+					"name" = "Trail",
+					"desc" = "Leave a trail behind."
+				)
+			)
+		)
+	)
+	data["upgrades"] = upgrades
+
+	return data
+
+/datum/mutation_datum/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+	switch(action)
+		if("purchase")
+			var/type = params["type"]
+			if(type == "Carapace")
```

Comment:
See if you can provide a type to the TGUI window and using that data apply the upgrade without having a huge if list